jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/PythonFunctionSignature.java

```diff
@@ -226,37 +239,51 @@ public boolean isFromArgumentSpec() {
+            case VIRTUAL:
+            case INTERFACE:
+            case CONSTRUCTOR:
+                return true;
+            default:
+                return false;
 
-        if (methodDescriptor.methodType == MethodDescriptor.MethodType.STATIC_AS_VIRTUAL) {
-            minParameters--;
-            maxParameters--;
-            startIndex = 1;
-        }
-        if (callParameters.length < minParameters || (extraPositionalArgumentsVariableIndex.isEmpty() &&
-                callParameters.length > maxParameters)) {
-            return false;
         }
+    }
 
-        for (int i = startIndex; i < Math.min(parameterTypes.length, callParameters.length); i++) {
-            if (extraPositionalArgumentsVariableIndex.isPresent() && i == extraPositionalArgumentsVariableIndex.get()) {
-                continue;
-            }
-            if (extraKeywordArgumentsVariableIndex.isPresent() && i == extraKeywordArgumentsVariableIndex.get()) {
-                continue;
-            }
+    public boolean isClassMethod() {
+        return methodDescriptor.methodType == MethodDescriptor.MethodType.CLASS;
+    }
 
-            PythonLikeType overloadParameterType = parameterTypes[i];
-            PythonLikeType callParameterType = callParameters[i];
+    public boolean isStaticMethod() {
+        return methodDescriptor.methodType == MethodDescriptor.MethodType.STATIC;
+    }
 
-            if (!callParameterType.isSubclassOf(overloadParameterType)) {
-                return false;
-            }
+    public boolean matchesParameters(PythonLikeType... callParameters) {
+        if (methodDescriptor.methodType == MethodDescriptor.MethodType.CLASS) {
+            List<PythonLikeType> actualCallParameters = new ArrayList<>();
+            actualCallParameters.add(BuiltinTypes.TYPE_TYPE);
+            actualCallParameters.addAll(List.of(callParameters));
+            return argumentSpec.verifyMatchesCallSignature(callParameters.length + 1, List.of(),
+                    actualCallParameters);
+        } else {
+            return argumentSpec.verifyMatchesCallSignature(callParameters.length, List.of(),
+                    Arrays.asList(callParameters));
+        }
+    }
+
+    public boolean matchesParameters(int positionalArgumentCount, List<String> keywordArgumentNameList,
+            List<PythonLikeType> callStackTypeList) {
+        if (methodDescriptor.methodType == MethodDescriptor.MethodType.CLASS) {
+            List<PythonLikeType> actualCallParameters = new ArrayList<>();
+            actualCallParameters.add(BuiltinTypes.TYPE_TYPE);
+            actualCallParameters.addAll(callStackTypeList);
+            return argumentSpec.verifyMatchesCallSignature(positionalArgumentCount + 1, keywordArgumentNameList,
+                    actualCallParameters);
+        } else {
+            return argumentSpec.verifyMatchesCallSignature(positionalArgumentCount, keywordArgumentNameList,
+                    callStackTypeList);
         }
```

Comment:
Looks to me as if these methods could delegate to a shared method. Their code does the same thing, just with different arguments.

Also, since `List.of()` results in an immutable empty list, `Collections.emptyList()` does a better job of showing that.